File: scripts/generate-data-openai.py
"""
Trivia App Data Generator using OpenAI API

This script generates JSON data for quizzes and their questions using the OpenAI API.
It reads quiz data from a CSV file and generates appropriate JSON files that match the existing format.

Usage: python generate-data-openai.py

Prerequisites:
- Python 3.6+
- OpenAI API key set as an environment variable OPENAI_API_KEY
- Required packages: openai, python-dotenv, os, json, datetime, re, random, pandas

Installation:
pip install openai python-dotenv pandas
"""

import os
import json
import datetime
import re
import time
import random
import pandas as pd
from openai import OpenAI

# Hardcoded API key directly in the code
OPENAI_API_KEY = "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"

# Initialize OpenAI client
client = OpenAI(api_key=OPENAI_API_KEY)

# Configuration
OUTPUT_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "generated")
TIMESTAMP = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
MODEL = "gpt-4-turbo"  # Default to GPT-4 Turbo
CSV_FILE = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "Quiz-data.csv")

# Create output directory if it doesn't exist
os.makedirs(OUTPUT_DIR, exist_ok=True)

def call_openai_api(system_prompt, user_prompt, temperature=0.7):
    """
    Call OpenAI API with system and user prompts
    
    Returns:
        str: The API response content
    """
    try:
        response = client.chat.completions.create(
            model=MODEL,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt}
            ],
            temperature=temperature
        )
        return response.choices[0].message.content
    except Exception as e:
        print(f"Error calling OpenAI API: {e}")
        raise

def read_quiz_data_from_csv():
    """
    Read quiz data from CSV file
    
    Returns:
        tuple: (category, subcategories_dict)
        where subcategories_dict is a dictionary mapping subcategory names to their IDs
    """
    try:
        df = pd.read_csv(CSV_FILE)
        if df.empty:
            raise ValueError("CSV file is empty")
            
        # Get unique category (should be 'Entertainment' for all rows)
        categories = df['Category ID'].unique()
        if len(categories) != 1:
            raise ValueError(f"Expected exactly one category, found {len(categories)}")
        category = categories[0]
        
        # Create dictionary of subcategories and their IDs
        subcategories_dict = {}
        for _, row in df.iterrows():
            quiz_name = row['Quiz'].replace(' Trivia', '')  # Remove 'Trivia' from the name
            subcategories_dict[quiz_name] = row['Subcategory ID']
            
        return category, subcategories_dict
    except Exception as e:
        print(f"Error reading CSV file: {e}")
        raise

def clean_json_string(json_str):
    """
    Clean a JSON string to handle common issues like control characters
    
    Args:
        json_str (str): The JSON string to clean
        
    Returns:
        str: The cleaned JSON string
    """
    # Remove any control characters
    json_str = re.sub(r'[\x00-\x1F\x7F]', '', json_str)
    
    # Handle any trailing commas in arrays or objects
    json_str = re.sub(r',\s*}', '}', json_str)
    json_str = re.sub(r',\s*]', ']', json_str)
    
    return json_str

def generate_quizzes_data(category, subcategories_dict):
    """
    Generate quizzes data for the given category and subcategories
    
    Args:
        category (str): The category name
        subcategories_dict (dict): Dictionary mapping subcategory names to their IDs
        
    Returns:
        dict: Quizzes data in the format matching quizzes.json
    """
    # Get category ID (lowercase with hyphens)
    category_id = category.lower().replace(" ", "-")
    
    # Always use "entertainment" as the categoryId for now
    # This can be modified if more categories are added
    category_id_fixed = "entertainment"
    
    system_prompt = """
You are a JSON data generator for a trivia application.
Your task is to create JSON data for quizzes based on the given category and subcategories.
Each quiz should have realistic and engaging details.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Always use the provided categoryId
3. The subcategoryId must be derived from the subcategory name (lowercase with hyphens)
4. The id field MUST be the subcategory name converted to lowercase with hyphens, and MUST end with "-trivia-questions-answers"
   - Example: For "Kids Movies", the id should be "kids-movies-trivia-questions-answers"
5. Create quiz names that EXACTLY match the subcategory name
6. Write detailed, interesting descriptions that accurately describe what the quiz is about
7. Include appropriate difficulty levels (Easy, Medium, or Hard)
8. Generate realistic values for questionCount, timeLimit, popularity, and attempts
9. Create relevant and specific tags for each quiz (3-4 tags)
10. Include a studyTopicId that is related to the quiz topic (like "oscar-awards" for an Oscars quiz)
11. Make sure the data is well-formatted JSON
"""

    example_json = {
        "quizzes": [
            {
                "id": "oscar-winners-trivia-questions-answers",
                "categoryId": "entertainment",
                "subcategoryId": "movies",
                "name": "Oscar Winners",
                "description": "Test your knowledge about Academy Award-winning movies throughout history.",
                "questionCount": 30,
                "timeLimit": 20,
                "difficulty": "Medium",
                "popularity": 4.8,
                "attempts": 24500,
                "studyTopicId": "oscar-awards",
                "tags": [
                    "Oscars",
                    "Best Picture",
                    "Award Winners"
                ]
            }
        ]
    }

    # Convert subcategories_dict to a list of names
    subcategory_names = list(subcategories_dict.keys())
    
    # Split into batches of 15 (or less for the last batch)
    batch_size = 15
    all_quizzes = {"quizzes": []}
    
    # Process subcategories in batches
    for i in range(0, len(subcategory_names), batch_size):
        batch = subcategory_names[i:i+batch_size]
        print(f"Processing batch {i//batch_size + 1} of {(len(subcategory_names) + batch_size - 1) // batch_size} ({len(batch)} quizzes)...")
        
        user_prompt = f"""
Generate quiz data for these inputs:
- Category: {category}
- Subcategories: {', '.join(batch)}

IMPORTANT FORMAT REQUIREMENTS:
1. Always use "{category_id_fixed}" as the categoryId for all quizzes
2. For each quiz, use the provided subcategoryId from the dictionary
3. The id field MUST follow this format: [subcategory-name-in-lowercase-with-hyphens]-trivia-questions-answers
4. The quiz name MUST be EXACTLY the same as the subcategory name
5. Include a studyTopicId field that relates to the quiz topic

I need one quiz for each subcategory. Please make sure all data is realistic with engaging descriptions.

Use this JSON format:
{json.dumps(example_json, indent=2)}

Return only valid, well-formatted JSON with NO additional text or explanations.
"""

        response = call_openai_api(system_prompt, user_prompt)
        
        try:
            # Clean the response and parse as JSON
            cleaned_response = clean_json_string(response)
            batch_quizzes_data = json.loads(cleaned_response)
            
            # Add subcategory_id to each quiz in this batch
            for quiz in batch_quizzes_data["quizzes"]:
                # Try to find the matching subcategory
                for subcat_name, subcat_id in subcategories_dict.items():
                    if quiz["name"] == subcat_name:
                        quiz["subcategoryId"] = subcat_id
                        break
            
            # Add quizzes from this batch to the overall list
            all_quizzes["quizzes"].extend(batch_quizzes_data["quizzes"])
            
            # Add a small delay to avoid rate limiting
            time.sleep(2)
        except json.JSONDecodeError as e:
            print(f"Failed to parse quizzes data from OpenAI API response: {e}")
            print("Attempting to fix JSON manually...")
            
            # Try to extract just the quizzes array if the full JSON can't be parsed
            try:
                # Look for the quizzes array
                match = re.search(r'"quizzes"\s*:\s*\[(.*?)\]\s*}', response, re.DOTALL)
                if match:
                    # Try to parse just the quizzes array
                    quizzes_array = f'{{"quizzes": [{match.group(1)}]}}'
                    cleaned_quizzes = clean_json_string(quizzes_array)
                    batch_quizzes_data = json.loads(cleaned_quizzes)
                    
                    # Add quizzes from this batch to the overall list
                    all_quizzes["quizzes"].extend(batch_quizzes_data["quizzes"])
                    print(f"Successfully extracted {len(batch_quizzes_data['quizzes'])} quizzes from partial JSON")
                    
                    # Add a small delay to avoid rate limiting
                    time.sleep(2)
                    continue
            except Exception as inner_e:
                print(f"Failed to extract quizzes from partial JSON: {inner_e}")
            
            print("Response was:", response)
            print("Skipping this batch and continuing...")
            time.sleep(2)
            continue  # Skip this batch but continue with the next one
    
    return all_quizzes

def generate_questions_data(quizzes_data):
    """
    Generate questions data for the given quizzes
    
    Args:
        quizzes_data (dict): The quizzes data from which to generate questions
        
    Returns:
        dict: Questions data in the format matching questions.json
    """
    questions_data = {"quizzes": []}
    
    # For each quiz, generate questions
    for quiz in quizzes_data["quizzes"]:
        print(f"Generating questions for quiz: {quiz['name']}...")
        
        # Random number of questions between 15 and 20
        question_count = random.randint(15, 20)
        
        system_prompt = """
You are a quiz content creator specializing in creating trivia questions.
Your task is to generate a set of 15-20 questions for a specific quiz topic.
Each question should be well-researched, factually accurate, and engaging.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Create the EXACT number of questions specified in the prompt (between 15-20)
3. Each question should have 4 distinct options (all reasonable possibilities, not obvious wrong answers)
4. Mark the correct answer and provide a brief, educational explanation of why it's correct
5. Vary the difficulty levels (Easy, Medium, Hard)
6. Each question must have a unique ID in the format "q1", "q2", etc.
7. Make sure the data is well-formatted JSON
8. Always use the provided Category and Subcategory values from the input
9. The quiz title MUST be in the format: [Quiz Name] + "Trivia Questions and Answers"
   - Example: "Kids Movies Trivia Questions and Answers"
10. The quiz ID MUST be exactly the same as provided in the input
"""

        example_json = {
            "id": "oscar-winners-trivia-questions-answers",
            "title": "Oscar Winners Trivia Questions and Answers",
            "category": "Entertainment",
            "subcategory": "Movies",
            "description": "Test your knowledge about Academy Award-winning movies throughout history.",
            "difficulty": "Medium",
            "timeLimit": 20,
            "questionCount": 15,
            "popularity": 4.8,
            "attempts": 24500,
            "tags": ["Oscars", "Best Picture", "Award Winners"],
            "questions": [
                {
                    "id": "q1",
                    "text": "Which film won the Academy Award for Best Picture in 2020?",
                    "options": [
                        "1917",
                        "Parasite",
                        "Joker",
                        "Once Upon a Time in Hollywood"
                    ],
                    "correctAnswer": "Parasite",
                    "explanation": "Parasite, directed by Bong Joon-ho, became the first non-English language film to win the Academy Award for Best Picture in 2020.",
                    "difficulty": "Medium"
                },
                {
                    "id": "q2",
                    "text": "Who holds the record for most Academy Award wins for Best Director?",
                    "options": [
                        "Steven Spielberg",
                        "Martin Scorsese",
                        "John Ford",
                        "Francis Ford Coppola"
                    ],
                    "correctAnswer": "John Ford",
                    "explanation": "John Ford won the Academy Award for Best Director four times for 'The Informer' (1935), 'The Grapes of Wrath' (1940), 'How Green Was My Valley' (1941), and 'The Quiet Man' (1952).",
                    "difficulty": "Hard"
                }
            ]
        }

        user_prompt = f"""
Generate questions data for this quiz:
- Quiz ID: {quiz['id']}
- Quiz Title: {quiz['name']} Trivia Questions and Answers
- Category: {quiz.get('categoryId', 'Entertainment')}
- Subcategory: {quiz.get('subcategoryId', 'Movies')}
- Description: {quiz['description']}

IMPORTANT FORMAT REQUIREMENTS:
1. Always use "{quiz.get('categoryId', 'Entertainment').capitalize()}" as the category
2. Always use "{quiz.get('subcategoryId', 'Movies').capitalize()}" as the subcategory
3. Use the exact format shown in the example
4. IMPORTANT: The ID must be EXACTLY: {quiz['id']}
5. The title must be EXACTLY: "{quiz['name']} Trivia Questions and Answers"

Create {question_count} questions for this quiz. Follow these guidelines:
1. Questions should be factually accurate and cover different aspects of {quiz['name']}
2. Each question should have 4 reasonable options (not obviously wrong answers)
3. Include a brief explanation for each correct answer
4. Vary the difficulty levels (Easy, Medium, Hard)

Use this JSON format:
{json.dumps(example_json, indent=2)}

Return only valid, well-formatted JSON with NO additional text or explanations.
"""

        response = call_openai_api(system_prompt, user_prompt)
        
        try:
            # Clean the response and parse as JSON
            cleaned_response = clean_json_string(response)
            questions_quiz_data = json.loads(cleaned_response)
            questions_data["quizzes"].append(questions_quiz_data)
            
            # Add a small delay to avoid rate limiting
            time.sleep(1)
        except json.JSONDecodeError as e:
            print(f"Failed to parse questions data for quiz {quiz['name']} from OpenAI API response: {e}")
            print("Skipping this quiz and continuing...")
            time.sleep(1)
            continue  # Skip this quiz but continue with the next one
    
    return questions_data

def main():
    print("🔍 Trivia App Data Generator using OpenAI API")
    print("============================================")
    
    try:
        # Read quiz data from CSV
        print("\n📖 Reading quiz data from CSV file...")
        category, subcategories_dict = read_quiz_data_from_csv()
        
        print(f"\n✅ Read data from CSV:")
        print(f"- Category: {category}")
        print(f"- Number of subcategories: {len(subcategories_dict)}")
        
        if len(subcategories_dict) == 0:
            print("❌ No valid subcategories found in CSV. Exiting...")
            return
        
        print(f"\n📝 Generating data for {len(subcategories_dict)} subcategories using OpenAI API...")
        
        # Step 1: Generate quizzes data
        print("\n🎮 Generating quizzes data...")
        quizzes_data = generate_quizzes_data(category, subcategories_dict)
        print(f"✅ Generated data for {len(quizzes_data['quizzes'])} quizzes")
        
        # Step 2: Generate questions data
        print("\n❓ Generating questions data...")
        questions_data = generate_questions_data(quizzes_data)
        print(f"✅ Generated data for {len(questions_data['quizzes'])} sets of questions")
        
        # Write data to files
        quizzes_file_path = os.path.join(OUTPUT_DIR, f"quizzes_{TIMESTAMP}.json")
        questions_file_path = os.path.join(OUTPUT_DIR, f"questions_{TIMESTAMP}.json")
        
        with open(quizzes_file_path, 'w', encoding='utf-8') as f:
            json.dump(quizzes_data, f, indent=2)
        
        with open(questions_file_path, 'w', encoding='utf-8') as f:
            json.dump(questions_data, f, indent=2)
        
        print("\n✅ Data generation complete!")
        print("\nGenerated files:")
        print(f"1. {quizzes_file_path}")
        print(f"2. {questions_file_path}")
        
        print("\n📝 Next steps:")
        print("1. Review the generated files")
        print("2. Manually copy the data to your actual JSON files")
        print("3. Make any necessary adjustments to match your needs")
        
    except Exception as e:
        print(f"\n❌ Error: {e}")

if __name__ == "__main__":
    main() 

File: scripts/generate-data-openai.js
/**
 * Data Generator Script for Trivia App using OpenAI API
 * 
 * This script generates JSON data for subcategories, quizzes, and questions
 * using the OpenAI API. It creates high-quality data that follows the same 
 * structure as the existing JSON files.
 * 
 * Usage: node generate-data-openai.js
 * 
 * Prerequisites:
 * - Node.js installed
 * - OpenAI API key set as an environment variable OPENAI_API_KEY
 * - Required npm packages installed: openai, dotenv, fs, path, readline
 * 
 * Installation:
 * npm install openai dotenv
 */

require('dotenv').config();
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const OpenAI = require('openai');

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// Configuration
const OUTPUT_DIR = path.join(__dirname, '../data/generated');
const TIMESTAMP = new Date().toISOString().replace(/[:.]/g, '-');
const MODEL = 'gpt-4-turbo'; // You can change this to 'gpt-3.5-turbo' for lower costs

// Check if output directory exists, if not create it
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

// Create readline interface for user input
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

/**
 * Promisify readline question
 * @param {string} question - The question to ask
 * @returns {Promise<string>} - The user's answer
 */
function askQuestion(question) {
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      resolve(answer);
    });
  });
}

/**
 * Call OpenAI API with a system prompt and user prompt
 * @param {string} systemPrompt - Instructions for the model
 * @param {string} userPrompt - User input/query
 * @returns {Promise<string>} - The API response
 */
async function callOpenAI(systemPrompt, userPrompt) {
  try {
    const response = await openai.chat.completions.create({
      model: MODEL,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt }
      ],
      temperature: 0.7,
    });

    return response.choices[0].message.content;
  } catch (error) {
    console.error('Error calling OpenAI API:', error);
    throw error;
  }
}

/**
 * Generate subcategories data using OpenAI
 * @param {string} categoryId - The parent category ID
 * @param {string} categoryName - The parent category name
 * @param {Array<string>} subcategoryNames - List of subcategory names
 * @returns {Promise<Object>} - Subcategories data object
 */
async function generateSubcategoriesData(categoryId, categoryName, subcategoryNames) {
  const systemPrompt = `
You are a helpful assistant that generates JSON data for a trivia application. 
Your task is to create subcategory data for the given category and subcategory names.
Each subcategory should have realistic and interesting details.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Use the provided category ID and name
3. Create appropriate IDs by converting names to lowercase and replacing spaces with hyphens
4. Create engaging descriptions that accurately describe the subcategory
5. Use realistic image URLs from Unsplash
6. Include appropriate difficulty levels (Easy, Medium, or Hard)
7. Make sure the data is well-formatted JSON
`;

  const exampleJson = {
    "subcategories": [
      {
        "id": "movies",
        "categoryId": "entertainment",
        "name": "Movies",
        "description": "Test your knowledge of cinema and film history",
        "image": "https://images.unsplash.com/photo-1485846234645-a62644f84728?q=80&w=2059&auto=format&fit=crop",
        "totalQuizzes": 55,
        "difficulty": "Medium",
        "ageGroup": "12+"
      },
      {
        "id": "tv-shows",
        "categoryId": "entertainment",
        "name": "TV Shows",
        "description": "Challenge yourself with television trivia",
        "image": "https://images.unsplash.com/photo-1593784991095-a205069470b6?q=80&w=2070&auto=format&fit=crop",
        "totalQuizzes": 48,
        "difficulty": "Easy",
        "ageGroup": "10+"
      }
    ]
  };

  const userPrompt = `
Generate subcategory data for these inputs:
- Category ID: ${categoryId}
- Category Name: ${categoryName}
- Subcategory Names: ${subcategoryNames.join(', ')}

Use this JSON format:
${JSON.stringify(exampleJson, null, 2)}

Return only valid, well-formatted JSON with no additional text or explanations.
`;

  const response = await callOpenAI(systemPrompt, userPrompt);
  
  try {
    // Parse the response as JSON
    return JSON.parse(response);
  } catch (error) {
    console.error('Error parsing OpenAI subcategories response:', error);
    console.error('Response was:', response);
    throw new Error('Failed to parse subcategories data from OpenAI');
  }
}

/**
 * Generate study subcategories data using OpenAI
 * @param {string} mainTopicId - The parent main topic ID
 * @param {Array<string>} subcategoryNames - List of subcategory names
 * @returns {Promise<Object>} - Study subcategories data object
 */
async function generateStudySubcategoriesData(mainTopicId, subcategoryNames) {
  const systemPrompt = `
You are a helpful assistant that generates JSON data for a trivia study application. 
Your task is to create study subcategory data for the given main topic and subcategory names.
Each study subcategory should have realistic and educational details.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Use the provided main topic ID
3. Create appropriate IDs by converting names to lowercase and replacing spaces with hyphens
4. Create educational descriptions that accurately describe the study subcategory
5. Include detailed meta descriptions for SEO
6. Use realistic image URLs from Unsplash
7. Generate relevant popular tags for each subcategory
8. Make sure the data is well-formatted JSON
`;

  const exampleJson = {
    "subcategories": [
      {
        "id": "movies",
        "mainTopicId": "entertainment",
        "name": "Movies",
        "description": "Explore the world of cinema, including films, directors, awards, and film history",
        "metaDescription": "Study materials on film history, movie awards, iconic directors, and significant movements in cinema history.",
        "image": "https://images.unsplash.com/photo-1536440136628-849c177e76a1?q=80&w=2025&auto=format&fit=crop",
        "totalStudyMaterials": 5,
        "popularTags": ["Film History", "Academy Awards", "Directors", "Film Techniques"]
      },
      {
        "id": "music",
        "mainTopicId": "entertainment",
        "name": "Music",
        "description": "Discover the rich history of music, famous artists, and influential genres",
        "metaDescription": "Comprehensive study materials on music history, influential artists, genre evolution, and musical theory and techniques.",
        "image": "https://images.unsplash.com/photo-1511671782779-c97d3d27a1d4?q=80&w=2070&auto=format&fit=crop",
        "totalStudyMaterials": 4,
        "popularTags": ["Music History", "Famous Musicians", "Genres", "Music Theory"]
      }
    ]
  };

  const userPrompt = `
Generate study subcategory data for these inputs:
- Main Topic ID: ${mainTopicId}
- Subcategory Names: ${subcategoryNames.join(', ')}

Use this JSON format:
${JSON.stringify(exampleJson, null, 2)}

Return only valid, well-formatted JSON with no additional text or explanations.
`;

  const response = await callOpenAI(systemPrompt, userPrompt);
  
  try {
    // Parse the response as JSON
    return JSON.parse(response);
  } catch (error) {
    console.error('Error parsing OpenAI study subcategories response:', error);
    console.error('Response was:', response);
    throw new Error('Failed to parse study subcategories data from OpenAI');
  }
}

/**
 * Generate quizzes data using OpenAI
 * @param {string} categoryId - The parent category ID
 * @param {string} categoryName - The parent category name
 * @param {Array<Object>} subcategories - List of subcategory objects
 * @returns {Promise<Object>} - Quizzes data object
 */
async function generateQuizzesData(categoryId, categoryName, subcategories) {
  const systemPrompt = `
You are a helpful assistant that generates JSON data for a trivia application. 
Your task is to create quiz data for the given category and subcategories.
Each quiz should have realistic and engaging details that would interest users.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Use the provided category ID and subcategory IDs
3. Create appropriate quiz IDs by converting names to lowercase and replacing spaces with hyphens
4. Create engaging descriptions that accurately describe what the quiz is about
5. Include appropriate difficulty levels (Easy, Medium, or Hard)
6. Generate relevant and specific tags for each quiz
7. Make sure the data is well-formatted JSON
8. Create one quiz for each subcategory
`;

  const exampleJson = {
    "quizzes": [
      {
        "id": "oscar-winners",
        "categoryId": "entertainment",
        "subcategoryId": "movies",
        "name": "Oscar-Winning Films",
        "description": "Test your knowledge about Academy Award-winning movies throughout history.",
        "questionCount": 20,
        "timeLimit": 15,
        "difficulty": "Medium",
        "popularity": 4.8,
        "attempts": 24500,
        "tags": ["Oscars", "Best Picture", "Award Winners"]
      },
      {
        "id": "sitcoms-trivia",
        "categoryId": "entertainment",
        "subcategoryId": "tv-shows",
        "name": "Classic Sitcoms",
        "description": "Challenge yourself with questions about beloved situation comedies from television history.",
        "questionCount": 15,
        "timeLimit": 12,
        "difficulty": "Easy",
        "popularity": 4.6,
        "attempts": 18700,
        "tags": ["Sitcoms", "Comedy", "TV Classics"]
      }
    ]
  };

  const userPrompt = `
Generate quiz data for these inputs:
- Category ID: ${categoryId}
- Category Name: ${categoryName}
- Subcategories: ${JSON.stringify(subcategories.map(s => ({ id: s.id, name: s.name })))}

Use this JSON format:
${JSON.stringify(exampleJson, null, 2)}

Return only valid, well-formatted JSON with no additional text or explanations.
`;

  const response = await callOpenAI(systemPrompt, userPrompt);
  
  try {
    // Parse the response as JSON
    return JSON.parse(response);
  } catch (error) {
    console.error('Error parsing OpenAI quizzes response:', error);
    console.error('Response was:', response);
    throw new Error('Failed to parse quizzes data from OpenAI');
  }
}

/**
 * Generate questions data using OpenAI
 * @param {string} categoryName - The parent category name
 * @param {Array<Object>} quizzes - List of quiz objects with subcategory info
 * @returns {Promise<Object>} - Questions data object
 */
async function generateQuestionsData(categoryName, quizzes) {
  // We'll generate questions for each quiz separately to keep prompts smaller
  const questionsDataArray = [];
  
  for (const quiz of quizzes) {
    const systemPrompt = `
You are a quiz expert that creates realistic, challenging, and engaging trivia questions. 
Your task is to generate a set of 10 questions for a specific quiz category.
Each question should be well-researched, factually accurate, and interesting.

Follow these guidelines:
1. Generate JSON data in the exact format shown in the example
2. Create 10 questions for the quiz
3. Each question should have 4 options (A, B, C, D)
4. Mark the correct answer and provide a brief, educational explanation
5. Vary the difficulty levels (Easy, Medium, Hard)
6. Ensure questions are diverse and cover different aspects of the topic
7. Make sure the data is well-formatted JSON
`;

    const exampleJson = {
      "id": "oscar-winners",
      "title": "Oscar-Winning Films",
      "category": "Entertainment",
      "subcategory": "Movies",
      "description": "Test your knowledge about Academy Award-winning movies throughout history.",
      "difficulty": "Medium",
      "timeLimit": 15,
      "questionCount": 10,
      "popularity": 4.5,
      "attempts": 10000,
      "tags": ["Oscars", "Best Picture", "Award Winners"],
      "questions": [
        {
          "id": "q1",
          "text": "Which film won the Academy Award for Best Picture in 2020?",
          "options": [
            "1917",
            "Parasite",
            "Joker",
            "Once Upon a Time in Hollywood"
          ],
          "correctAnswer": "Parasite",
          "explanation": "Parasite, directed by Bong Joon-ho, became the first non-English language film to win the Academy Award for Best Picture in 2020.",
          "difficulty": "Medium"
        },
        {
          "id": "q2",
          "text": "Who holds the record for most Academy Award wins for Best Director?",
          "options": [
            "Steven Spielberg",
            "Martin Scorsese",
            "John Ford",
            "Francis Ford Coppola"
          ],
          "correctAnswer": "John Ford",
          "explanation": "John Ford won the Academy Award for Best Director four times for 'The Informer' (1935), 'The Grapes of Wrath' (1940), 'How Green Was My Valley' (1941), and 'The Quiet Man' (1952).",
          "difficulty": "Hard"
        }
      ]
    };

    const userPrompt = `
Generate questions data for this quiz:
- Quiz ID: ${quiz.id}
- Quiz Title: ${quiz.name}
- Category: ${categoryName}
- Subcategory: ${quiz.subcategoryName}
- Description: ${quiz.description}

Create 10 realistic, engaging, and factually accurate questions with 4 options each.
Ensure questions cover different aspects of the topic and vary in difficulty.

Use this JSON format:
${JSON.stringify(exampleJson, null, 2)}

Return only valid, well-formatted JSON with no additional text or explanations.
`;

    const response = await callOpenAI(systemPrompt, userPrompt);
    
    try {
      // Parse the response as JSON
      const questionsData = JSON.parse(response);
      questionsDataArray.push(questionsData);
    } catch (error) {
      console.error(`Error parsing OpenAI questions response for quiz ${quiz.id}:`, error);
      console.error('Response was:', response);
      throw new Error(`Failed to parse questions data for quiz ${quiz.id} from OpenAI`);
    }
  }
  
  return { quizzes: questionsDataArray };
}

/**
 * Main function to generate all data
 */
async function generateData() {
  try {
    console.log('🔍 Trivia App Data Generator using OpenAI');
    console.log('========================================');
    
    // Check for API key
    if (!process.env.OPENAI_API_KEY) {
      console.error('❌ OPENAI_API_KEY environment variable not set');
      console.log('Please set your OpenAI API key using:');
      console.log('- Create a .env file in the project root with OPENAI_API_KEY=your_api_key');
      console.log('- Or set it in your environment before running the script');
      rl.close();
      return;
    }
    
    // Ask for category ID and name
    const categoryId = await askQuestion('Enter category ID (e.g., "entertainment"): ');
    const categoryName = await askQuestion('Enter category name (e.g., "Entertainment"): ');
    
    // Ask for subcategory information
    console.log('\nNow enter up to 5 subcategories for this category.');
    console.log('Example format: "Action Movies, Classic Movies, TV Shows, Sci-Fi, Music"\n');
    
    const subcategoriesInput = await askQuestion('Enter subcategories (comma-separated): ');
    const subcategoriesList = subcategoriesInput.split(',').map(item => item.trim()).filter(Boolean);
    
    // Limit to 5 subcategories
    const subcategories = subcategoriesList.slice(0, 5);
    
    if (subcategories.length === 0) {
      console.log('❌ No valid subcategories provided. Exiting...');
      rl.close();
      return;
    }
    
    console.log(`\n✅ Generating data for ${subcategories.length} subcategories using OpenAI API...`);
    
    // Step 1: Generate subcategories data
    console.log('\n📝 Generating subcategories data...');
    const subcategoriesData = await generateSubcategoriesData(categoryId, categoryName, subcategories);
    console.log(`✅ Generated data for ${subcategoriesData.subcategories.length} subcategories`);
    
    // Step 2: Generate study subcategories data
    console.log('\n📝 Generating study subcategories data...');
    const studySubcategoriesData = await generateStudySubcategoriesData(categoryId, subcategories);
    console.log(`✅ Generated data for ${studySubcategoriesData.subcategories.length} study subcategories`);
    
    // Step 3: Generate quizzes data
    console.log('\n📝 Generating quizzes data...');
    const quizzesData = await generateQuizzesData(categoryId, categoryName, subcategoriesData.subcategories);
    console.log(`✅ Generated data for ${quizzesData.quizzes.length} quizzes`);
    
    // Step 4: Generate questions data
    console.log('\n📝 Generating questions data...');
    // Add subcategory names to quiz objects for better context in question generation
    const enrichedQuizzes = quizzesData.quizzes.map(quiz => {
      const subcategory = subcategoriesData.subcategories.find(s => s.id === quiz.subcategoryId);
      return {
        ...quiz,
        subcategoryName: subcategory ? subcategory.name : 'Unknown'
      };
    });
    
    const questionsData = await generateQuestionsData(categoryName, enrichedQuizzes);
    console.log(`✅ Generated data for ${questionsData.quizzes.length} sets of questions`);
    
    // Write data to files
    const subcategoriesFilePath = path.join(OUTPUT_DIR, `subcategories_${TIMESTAMP}.json`);
    const studySubcategoriesFilePath = path.join(OUTPUT_DIR, `study_subcategories_${TIMESTAMP}.json`);
    const quizzesFilePath = path.join(OUTPUT_DIR, `quizzes_${TIMESTAMP}.json`);
    const questionsFilePath = path.join(OUTPUT_DIR, `questions_${TIMESTAMP}.json`);
    
    fs.writeFileSync(subcategoriesFilePath, JSON.stringify(subcategoriesData, null, 2));
    fs.writeFileSync(studySubcategoriesFilePath, JSON.stringify(studySubcategoriesData, null, 2));
    fs.writeFileSync(quizzesFilePath, JSON.stringify(quizzesData, null, 2));
    fs.writeFileSync(questionsFilePath, JSON.stringify(questionsData, null, 2));
    
    console.log('\n✅ Data generation complete!');
    console.log('\nGenerated files:');
    console.log(`1. ${subcategoriesFilePath}`);
    console.log(`2. ${studySubcategoriesFilePath}`);
    console.log(`3. ${quizzesFilePath}`);
    console.log(`4. ${questionsFilePath}`);
    
    console.log('\n📝 Next steps:');
    console.log('1. Review the generated files');
    console.log('2. Manually copy the data to your actual JSON files');
    console.log('3. Make any necessary adjustments to match your needs');
    
    rl.close();
  } catch (error) {
    console.error('❌ Error generating data:', error);
    rl.close();
  }
}

// Start the data generation process
generateData(); 